### ISSUE TITLE
Bug: [no-useless-assignment] false positive when returned after a branch where one clause modifies the variable and another diverges

### ISSUE THREAD

**Issue by gwennlbh** [eslint/eslint#20491]
### Environment

ESLint Playground

### What parser are you using?

Default (Espree)

### Playground Code
```js
function props(parameters) {
	let content = '';

  	if (typeof parameters === 'string') {
      content = parameters;
    } else {
  		return { };
	}

	return { content };
}
props()

```

### ESLint Configuration 
<details> 
 <summary> Tap to see full config </summary> 

```js
import { defineConfig } from "eslint/config";

export default defineConfig([
    {
        "rules": {
            "constructor-super": [
                "error"
            ],
            "for-direction": [
                "error"
            ],
            "getter-return": [
                "error"
            ],
            "no-async-promise-executor": [
                "error"
            ],
            "no-case-declarations": [
                "error"
            ],
            "no-class-assign": [
                "error"
            ],
            "no-compare-neg-zero": [
                "error"
            ],
            "no-cond-assign": [
                "error"
            ],
            "no-const-assign": [
                "error"
            ],
            "no-constant-binary-expression": [
                "error"
            ],
            "no-constant-condition": [
                "error"
            ],
            "no-control-regex": [
                "error"
            ],
            "no-debugger": [
                "error"
            ],
            "no-delete-var": [
                "error"
            ],
            "no-dupe-args": [
                "error"
            ],
            "no-dupe-class-members": [
                "error"
            ],
            "no-dupe-else-if": [
                "error"
            ],
            "no-dupe-keys": [
                "error"
            ],
            "no-duplicate-case": [
                "error"
            ],
            "no-empty": [
                "error"
            ],
            "no-empty-character-class": [
                "error"
            ],
            "no-empty-pattern": [
                "error"
            ],
            "no-empty-static-block": [
                "error"
            ],
            "no-ex-assign": [
                "error"
            ],
            "no-extra-boolean-cast": [
                "error"
            ],
            "no-fallthrough": [
                "error"
            ],
            "no-func-assign": [
                "error"
            ],
            "no-global-assign": [
                "error"
            ],
            "no-import-assign": [
                "error"
            ],
            "no-invalid-regexp": [
                "error"
            ],
            "no-irregular-whitespace": [
                "error"
            ],
            "no-loss-of-precision": [
                "error"
            ],
            "no-misleading-character-class": [
                "error"
            ],
            "no-new-native-nonconstructor": [
                "error"
            ],
            "no-nonoctal-decimal-escape": [
                "error"
            ],
            "no-obj-calls": [
                "error"
            ],
            "no-octal": [
                "error"
            ],
            "no-prototype-builtins": [
                "error"
            ],
            "no-redeclare": [
                "error"
            ],
            "no-regex-spaces": [
                "error"
            ],
            "no-self-assign": [
                "error"
            ],
            "no-setter-return": [
                "error"
            ],
            "no-shadow-restricted-names": [
                "error"
            ],
            "no-sparse-arrays": [
                "error"
            ],
            "no-this-before-super": [
                "error"
            ],
            "no-unassigned-vars": [
                "error"
            ],
            "no-undef": [
                "error"
            ],
            "no-unexpected-multiline": [
                "error"
            ],
            "no-unreachable": [
                "error"
            ],
            "no-unsafe-finally": [
                "error"
            ],
            "no-unsafe-negation": [
                "error"
            ],
            "no-unsafe-optional-chaining": [
                "error"
            ],
            "no-unused-labels": [
                "error"
            ],
            "no-unused-private-class-members": [
                "error"
            ],
            "no-unused-vars": [
                "error"
            ],
            "no-useless-assignment": [
                "error"
            ],
            "no-useless-backreference": [
                "error"
            ],
            "no-useless-catch": [
                "error"
            ],
            "no-useless-escape": [
                "error"
            ],
            "no-with": [
                "error"
            ],
            "preserve-caught-error": [
                "error"
            ],
            "require-yield": [
                "error"
            ],
            "use-isnan": [
                "error"
            ],
            "valid-typeof": [
                "error"
            ]
        }
    }
]);
```
 </details>

### What did you expect to happen?

No errors

### What actually happened?

```
2:6 This assigned value is not used in subsequent statements. (no-useless-assignment)
```

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiZnVuY3Rpb24gcHJvcHMocGFyYW1ldGVycykge1xuXHRsZXQgY29udGVudCA9ICcnO1xuXG4gIFx0aWYgKHR5cGVvZiBwYXJhbWV0ZXJzID09PSAnc3RyaW5nJykge1xuICAgICAgY29udGVudCA9IHBhcmFtZXRlcnM7XG4gICAgfSBlbHNlIHtcbiAgXHRcdHJldHVybiB7IH07XG5cdH1cblxuXHRyZXR1cm4geyBjb250ZW50IH07XG59XG5wcm9wcygpXG4iLCJvcHRpb25zIjp7InJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjpbImVycm9yIl0sImZvci1kaXJlY3Rpb24iOlsiZXJyb3IiXSwiZ2V0dGVyLXJldHVybiI6WyJlcnJvciJdLCJuby1hc3luYy1wcm9taXNlLWV4ZWN1dG9yIjpbImVycm9yIl0sIm5vLWNhc2UtZGVjbGFyYXRpb25zIjpbImVycm9yIl0sIm5vLWNsYXNzLWFzc2lnbiI6WyJlcnJvciJdLCJuby1jb21wYXJlLW5lZy16ZXJvIjpbImVycm9yIl0sIm5vLWNvbmQtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbnN0LWFzc2lnbiI6WyJlcnJvciJdLCJuby1jb25zdGFudC1iaW5hcnktZXhwcmVzc2lvbiI6WyJlcnJvciJdLCJuby1jb25zdGFudC1jb25kaXRpb24iOlsiZXJyb3IiXSwibm8tY29udHJvbC1yZWdleCI6WyJlcnJvciJdLCJuby1kZWJ1Z2dlciI6WyJlcnJvciJdLCJuby1kZWxldGUtdmFyIjpbImVycm9yIl0sIm5vLWR1cGUtYXJncyI6WyJlcnJvciJdLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOlsiZXJyb3IiXSwibm8tZHVwZS1lbHNlLWlmIjpbImVycm9yIl0sIm5vLWR1cGUta2V5cyI6WyJlcnJvciJdLCJuby1kdXBsaWNhdGUtY2FzZSI6WyJlcnJvciJdLCJuby1lbXB0eSI6WyJlcnJvciJdLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOlsiZXJyb3IiXSwibm8tZW1wdHktcGF0dGVybiI6WyJlcnJvciJdLCJuby1lbXB0eS1zdGF0aWMtYmxvY2siOlsiZXJyb3IiXSwibm8tZXgtYXNzaWduIjpbImVycm9yIl0sIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6WyJlcnJvciJdLCJuby1mYWxsdGhyb3VnaCI6WyJlcnJvciJdLCJuby1mdW5jLWFzc2lnbiI6WyJlcnJvciJdLCJuby1nbG9iYWwtYXNzaWduIjpbImVycm9yIl0sIm5vLWltcG9ydC1hc3NpZ24iOlsiZXJyb3IiXSwibm8taW52YWxpZC1yZWdleHAiOlsiZXJyb3IiXSwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOlsiZXJyb3IiXSwibm8tbG9zcy1vZi1wcmVjaXNpb24iOlsiZXJyb3IiXSwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOlsiZXJyb3IiXSwibm8tbmV3LW5hdGl2ZS1ub25jb25zdHJ1Y3RvciI6WyJlcnJvciJdLCJuby1ub25vY3RhbC1kZWNpbWFsLWVzY2FwZSI6WyJlcnJvciJdLCJuby1vYmotY2FsbHMiOlsiZXJyb3IiXSwibm8tb2N0YWwiOlsiZXJyb3IiXSwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjpbImVycm9yIl0sIm5vLXJlZGVjbGFyZSI6WyJlcnJvciJdLCJuby1yZWdleC1zcGFjZXMiOlsiZXJyb3IiXSwibm8tc2VsZi1hc3NpZ24iOlsiZXJyb3IiXSwibm8tc2V0dGVyLXJldHVybiI6WyJlcnJvciJdLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6WyJlcnJvciJdLCJuby1zcGFyc2UtYXJyYXlzIjpbImVycm9yIl0sIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjpbImVycm9yIl0sIm5vLXVuYXNzaWduZWQtdmFycyI6WyJlcnJvciJdLCJuby11bmRlZiI6WyJlcnJvciJdLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6WyJlcnJvciJdLCJuby11bnJlYWNoYWJsZSI6WyJlcnJvciJdLCJuby11bnNhZmUtZmluYWxseSI6WyJlcnJvciJdLCJuby11bnNhZmUtbmVnYXRpb24iOlsiZXJyb3IiXSwibm8tdW5zYWZlLW9wdGlvbmFsLWNoYWluaW5nIjpbImVycm9yIl0sIm5vLXVudXNlZC1sYWJlbHMiOlsiZXJyb3IiXSwibm8tdW51c2VkLXByaXZhdGUtY2xhc3MtbWVtYmVycyI6WyJlcnJvciJdLCJuby11bnVzZWQtdmFycyI6WyJlcnJvciJdLCJuby11c2VsZXNzLWFzc2lnbm1lbnQiOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1iYWNrcmVmZXJlbmNlIjpbImVycm9yIl0sIm5vLXVzZWxlc3MtY2F0Y2giOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1lc2NhcGUiOlsiZXJyb3IiXSwibm8td2l0aCI6WyJlcnJvciJdLCJwcmVzZXJ2ZS1jYXVnaHQtZXJyb3IiOlsiZXJyb3IiXSwicmVxdWlyZS15aWVsZCI6WyJlcnJvciJdLCJ1c2UtaXNuYW4iOlsiZXJyb3IiXSwidmFsaWQtdHlwZW9mIjpbImVycm9yIl19LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
This works as intended because the `''` value assigned on initialization is never used. If `typeof parameters === 'string'` is `true`, it's overwritten with `parameters`. Else, the function exits due to the `return` statement so it doesn't reach `return { content };` where the value would be used. Overall, replacing `let content = '';` with `let content;` doesn't change the behavior of the function, so it's a useless assignment.

Note: I also checked the same example with ESLint v9.39.2, and the result is the same so this isn't a change in the behavior of this rule that might have been introduced in ESLint v10.0.0.

**Comment by gwennlbh** [comment]
oh yeah sorry. the rule *does* break lints when upgrading to v10 though, but due to other false positives in Svelte files (for example, when declaring a bindable prop with `let { a = $bindable() } = $props()`)
